### PR TITLE
Add support for consensus model V2

### PR DESCRIPTION
**Pull request by SebastianMarian**
* Added support for consensus model V2
* Added support for adding processed header hash into consensus messages
* Added support for generating bitmap, based on processed headers hashes received from validators in signature subround
* Added activation flag for consensus model V2

## Reasoning behind the pull request
- 
- 
- 
  
## Proposed changes
- 
- 
- 

## Testing procedure
- 
- 
- 

## Pre-requisites

Based on the [Contributing Guidelines](https://github.com/multiversx/mx-chain-go/blob/master/.github/CONTRIBUTING.md#branches-management) the PR author and the reviewers must check the following requirements are met:
- was the PR targeted to the correct branch?
- if this is a larger feature that probably needs more than one PR, is there a `feat` branch created?
- if this is a `feat` branch merging, do all satellite projects have a proper tag inside `go.mod`?

### CODE REVIEW

File: consensus/message.proto [hunk 1]
```diff
@@ -9,18 +9,19 @@ import "github.com/gogo/protobuf/gogoproto/gogo.proto";
...
-	int64    MsgType            = 7;
-	int64    RoundIndex         = 8;
-	bytes    ChainID            = 9;
-	bytes    PubKeysBitmap      = 10;
-	bytes    AggregateSignature = 11;
-	bytes    LeaderSignature    = 12;
-	bytes    OriginatorPid      = 13;
-	bytes    InvalidSigners     = 14;
+	bytes    HeaderHash						= 1;
+	bytes    SignatureShare				= 2;
```
Comment (sasurobert) [inline consensus/message.proto:13]: indent

Comment (SebastianMarian) [reply consensus/message.proto:13]: Done

File: consensus/spos/bls/blsSubroundsFactory.go [hunk 2]
```diff
@@ -161,6 +177,37 @@ func (fct *factory) generateStartRoundSubround() error {
 }
 
 func (fct *factory) generateBlockSubround() error {
+	subroundBlockInstance, err := fct.generateBlockSubroundV1()
```
[the remark this answers is not in the document]
Comment (SebastianMarian) [reply consensus/spos/bls/blsSubroundsFactory.go:180]: Refactored

File: consensus/spos/bls/subroundEndRound.go [hunk 3]
```diff
@@ -44,12 +44,14 @@ func NewSubroundEndRound(
 		baseSubround,
 		processingThresholdPercentage,
 		displayStatistics,
-		appStatusHandler,
+		baseSubround.AppStatusHandler(),
```
[the remark this answers is not in the document]
Comment (SebastianMarian) [reply consensus/spos/bls/subroundEndRound.go:47]: Actually this code should be the same also in the sovereign. I found that the AppStatusHandler was already set in baseSubround and not needed to send it as a parameter in each NewSubround... . This is an old code which could be simplified.

File: consensus/spos/bls/subroundSignature.go [hunk 4]
```diff
@@ -15,32 +15,30 @@ import (
 type subroundSignature struct {
 	*spos.Subround
 
-	appStatusHandler core.AppStatusHandler
+	appStatusHandler     core.AppStatusHandler
```
[the remark this answers is not in the document]
Comment (SebastianMarian) [reply consensus/spos/bls/subroundSignature.go:18]: Actually this code should be the same also in the sovereign. I found that the AppStatusHandler was already set in baseSubround and not needed to send it as a parameter in each NewSubround... . This is an old code which could be simplified. The check for nil for AppStatusHandler is already done in generateEndRoundSubround. So I will add some unit tests for the small methods extracted here, to easily handle the consensus model v2 approach, after the activation.

File: consensus/spos/consensusMessageValidator.go [hunk 5]
```diff
@@ -195,15 +208,32 @@ func (cmv *consensusMessageValidator) checkConsensusMessageValidity(cnsMsg *cons
...
+	if !cmv.enableEpochHandler.IsConsensusModelV2Enabled() {
+		return true
+	}
+
+	msgType := consensus.MessageType(cnsMsg.MsgType)
+	isMessageWithBlockBody := cmv.consensusService.IsMessageWithBlockBody(msgType)
+	isMessageWithBlockHeader := cmv.consensusService.IsMessageWithBlockHeader(msgType)
+	isMessageWithBlockBodyAndHeader := cmv.consensusService.IsMessageWithBlockBodyAndHeader(msgType)
+
+	if isMessageWithBlockBody || isMessageWithBlockHeader || isMessageWithBlockBodyAndHeader {
```
[the remark this answers is not in the document]
Comment (SebastianMarian) [reply consensus/spos/consensusMessageValidator.go:232]: Yep, isProcessedHeaderHashSizeValid is just one step in the whole checks done in checkConsensusMessageValidity. We need to have this field set on nil in all the consensus messages from subround Block, and not nil + equal with header hash size, in messages from subround Signature and EndRound

File: consensus/spos/bls/subroundEndRound.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (SebastianMarian) [reply consensus/spos/bls/subroundEndRound.go:47]: Added unit tests

File: consensus/spos/bls/subroundSignature.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (SebastianMarian) [reply consensus/spos/bls/subroundSignature.go:18]: Added unit tests

File: consensus/spos/worker_test.go [hunk 6]
```diff
@@ -149,7 +150,7 @@ func initRoundHandlerMock() *mock.RoundHandlerMock {
 	}
 }
 
-func TestWorker_NewWorkerConsensusServiceNilShouldFail(t *testing.T) {
+func TestWorker_NewWorkerNilConsensusServiceShouldFail(t *testing.T) {
```
[the remark this answers is not in the document]
Comment (SebastianMarian) [reply consensus/spos/worker_test.go:153]: Renamed

File: consensus/spos/consensusState.go [hunk 7]
```diff
@@ -296,26 +322,48 @@ func (cns *ConsensusState) CanProcessReceivedMessage(cnsDta *consensus.Message,
...
+	return bitmap
+}
+
+// GenerateBitmapForHash method generates a bitmap, for a given subround and a processed header hash,
+// in which each node will be marked with 1 if its job has been done
+func (cns *ConsensusState) GenerateBitmapForHash(subroundId int, hash []byte) []byte {
+	bitmap := cns.createEmptyBitmap(len(cns.ConsensusGroup()))
+
+	indexes, _ := cns.GetProcessedHeaderHashIndexes(hash)
+	for _, i := range indexes {
```
[the remark this answers is not in the document]
Comment (SebastianMarian) [reply consensus/spos/consensusState.go:365]: Done

**Review by sasurobert — APPROVED** [review]

**Review by iulianpascalau — APPROVED** [review]

**Pull request merged**